src/main/java/org/kuali/kfs/module/ar/businessobject/TimeBasedBillingPeriod.java

```diff
@@ -0,0 +1,125 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.kuali.kfs.module.ar.businessobject;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kuali.kfs.coa.businessobject.AccountingPeriod;
+import org.kuali.kfs.coa.service.AccountingPeriodService;
+import org.kuali.kfs.module.ar.ArConstants;
+
+import java.sql.Date;
+
+public class TimeBasedBillingPeriod extends BillingPeriod {
+    
+    /*
+     * CU Customization
+     */
+    private static final Logger LOG = LogManager.getLogger(TimeBasedBillingPeriod.class);
+    
+    public TimeBasedBillingPeriod(ArConstants.BillingFrequencyValues billingFrequency, Date awardStartDate, Date currentDate, Date lastBilledDate, AccountingPeriodService accountingPeriodService) {
+        super(billingFrequency, awardStartDate, currentDate, lastBilledDate, accountingPeriodService);
+    }
+
+    @Override
+    protected Date determineEndDateByFrequency() {
+        final AccountingPeriod accountingPeriod = findPreviousAccountingPeriod(currentDate);
+        return accountingPeriod.getUniversityFiscalPeriodEndDate();
+    }
+
+    protected Integer calculatePreviousPeriodByFrequency(Integer currentAccountingPeriodCode, int periodsInBillingFrequency) {
+        Integer previousAccountingPeriodCode;
+        final int subAmt = (currentAccountingPeriodCode % periodsInBillingFrequency) == 0 ? periodsInBillingFrequency : currentAccountingPeriodCode % periodsInBillingFrequency;
+
+        previousAccountingPeriodCode = currentAccountingPeriodCode - subAmt;
+        return previousAccountingPeriodCode;
+    }
+
+    @Override
+    protected boolean canThisBeBilledByBillingFrequency() {
+        if (ArConstants.BillingFrequencyValues.ANNUALLY.equals(billingFrequency) && accountingPeriodService.getByDate(lastBilledDate).getUniversityFiscalYear() >= accountingPeriodService.getByDate(currentDate).getUniversityFiscalYear()) {
+            return false;
+        } else if (StringUtils.equals(findPreviousAccountingPeriod(currentDate).getUniversityFiscalPeriodCode(), findPreviousAccountingPeriod(lastBilledDate).getUniversityFiscalPeriodCode()) &&
+            accountingPeriodService.getByDate(lastBilledDate).getUniversityFiscalYear().equals(accountingPeriodService.getByDate(currentDate).getUniversityFiscalYear())) {
+            return false;
+        }
+
+        return true;
+    }
+    
+    /*
+     * CU Customization
+     */
+    @Override
+    protected Date determineStartDateByFrequency() {
+        if (lastBilledDate == null) {
+            LOG.error("determineStartDateByFrequency, unable to calculate start date by frequency when the last billed date iis NULL");
```

Comment:
there is a spelling error in the logged message: "iis NULL" should be "is NULL"

Reply:
fixed